core/common/src/main/java/alluxio/resource/LockResource.java

```diff
@@ -35,24 +39,39 @@
    * @param lock the lock to acquire
    */
   public LockResource(Lock lock) {
-    this(lock, true);
+    this(lock, true, false);
   }
 
   /**
    * Creates a new instance of {@link LockResource} using the given lock.
    *
+   * This method may use the {@link Lock#tryLock()} method to gain ownership of the locks. The
+   * reason one might want to use this is to avoid the fairness heuristics within the
+   * {@link java.util.concurrent.locks.ReentrantReadWriteLock}'s NonFairSync which may block reader
+   * threads if a writer if the first in the queue.
+   *
    * @param lock the lock to acquire
    * @param acquireLock whether to lock the lock
+   * @param useTryLock whether or not use to {@link Lock#tryLock()}
    */
-  public LockResource(Lock lock, boolean acquireLock) {
+  public LockResource(Lock lock, boolean acquireLock, boolean useTryLock) {
     mLock = lock;
     if (acquireLock) {
-      mLock.lock();
+      if (useTryLock) {
+        while (!mLock.tryLock()) { // returns immediately
+          // The reason we don't use #tryLock(int, TimeUnit) here is because we found there is a bug
+          // somewhere in the internal accounting of the ReentrantRWLock that, even though all
+          // threads had released the lock, that a final thread would never be able to acquire it.
+          LockSupport.parkNanos(10000);
```

Comment:
Should this be a parameter? Also, do we want to loop indefinitely, or throw some sort of exception when it was looping for too long? I'd imagine throwing an exception would prevent possible deadlocks or long delays, but failures would be less-user friendly. What do you think?

Reply:
The current implementation before `tryLock` will also hang indefinitely, so I felt it was okay to not have any timeout for the `tryLock` code path.